File: Solutions/680-ValidPalindromeII/ValidPalindromeII.kt
class Solution {
    fun validPalindrome(s: String): Boolean {
        var left = 0
        var right = s.length - 1

        while (left <= right) {
            if (s[left] != s[right]) {
                return isValidPalindrome(s, left + 1, right) || isValidPalindrome(s, left, right - 1)
            }
            left++
            right--
        }

        return true
    }

    fun isValidPalindrome(
        s: String,
        left: Int,
        right: Int,
    ): Boolean {
        var leftPointer = left
        var rightPointer = right

        while (leftPointer <= rightPointer) {
            if (s[left] != s[right]) {
                return false
            }
            leftPointer++
            rightPointer--
        }

        return true
    }
}

File: Solutions/All/680-ValidPalindromeII/ValidPalindromeII.ts
function validPalindrome(s: string): boolean {
  return isValid(s, 0, s.length - 1, true);
}

function isValid(
  s: string,
  left: number,
  right: number,
  removalRemaining: boolean,
) {
  // Base case: if left exceeds right, or they are equal then input is a palindrome
  if (left >= right) return true;

  // Both
  if (s[left] === s[right]) {
    return isValid(s, left + 1, right - 1, removalRemaining);
  } else {
    // Check if a removal already happened. If so, return false.
    if (!removalRemaining) return false;

    // Check if a palidrome exists after removing from the left and the right, and update removal remaining to false.
    removalRemaining = false;
    return (
      isValid(s, left + 1, right, removalRemaining) ||
      isValid(s, left, right - 1, removalRemaining)
    );
  }
}
